packages/core/test/new-api/reconciliation/reconciler.ts

```diff
@@ -0,0 +1,244 @@
+        status: ExecutionStatus.STARTED,
+      },
+    });
+
+    assert.deepStrictEqual(reconiliationResult.reconciliationFailures, [
+      {
+        futureId: "Module1:Example",
+        failure:
+          "Future with id Module1:Example has changed from NAMED_CONTRACT_DEPLOYMENT to NAMED_LIBRARY_DEPLOYMENT",
+      },
+    ]);
+  });
+
+  describe("dependencies", () => {
+    it("should reconcile unchanged dependencies", () => {
+      const moduleDefinition = defineModule("Module", (m) => {
+        const contract1 = m.contract("Contract1");
+        const contract2 = m.contract("Contract2");
+        const contract3 = m.contract("Contract3", [], {
+          after: [contract1, contract2],
+        });
+
+        return { contract1, contract2, contract3 };
+      });
+
+      assertSuccessReconciliation(moduleDefinition, {
+        "Module:Contract1": {
+          ...exampleDeploymentState,
+          futureType: FutureType.NAMED_CONTRACT_DEPLOYMENT,
+          status: ExecutionStatus.SUCCESS,
+          contractName: "Contract1",
+        },
+        "Module:Contract2": {
+          ...exampleDeploymentState,
+          futureType: FutureType.NAMED_CONTRACT_DEPLOYMENT,
+          status: ExecutionStatus.SUCCESS,
+          contractName: "Contract2",
+        },
+        "Module:Contract3": {
+          ...exampleDeploymentState,
+          futureType: FutureType.NAMED_CONTRACT_DEPLOYMENT,
+          status: ExecutionStatus.STARTED,
+          dependencies: new Set(["Module:Contract2", "Module:Contract2"]),
+          contractName: "Contract3",
+        },
+      });
+    });
+
+    it("should reconcile the reduction of dependencies", () => {
+      const moduleDefinition = defineModule("Module", (m) => {
+        const contract1 = m.contract("Contract1");
+        const contract2 = m.contract("Contract2");
+        const contract3 = m.contract("Contract3", [], {
+          after: [contract2], // one less dep than execution state
+        });
+
+        return { contract1, contract2, contract3 };
+      });
+
+      assertSuccessReconciliation(moduleDefinition, {
+        "Module:Contract1": {
+          ...exampleDeploymentState,
+          futureType: FutureType.NAMED_CONTRACT_DEPLOYMENT,
+          status: ExecutionStatus.SUCCESS,
+          contractName: "Contract1",
+        },
+        "Module:Contract2": {
+          ...exampleDeploymentState,
+          futureType: FutureType.NAMED_CONTRACT_DEPLOYMENT,
+          status: ExecutionStatus.SUCCESS,
+          contractName: "Contract2",
+        },
+        "Module:Contract3": {
+          ...exampleDeploymentState,
+          futureType: FutureType.NAMED_CONTRACT_DEPLOYMENT,
+          status: ExecutionStatus.SUCCESS,
+          dependencies: new Set(["Module:Contract1", "Module:Contract2"]),
+          contractName: "Contract3",
+        },
+      });
+    });
+
+    it("should reconcile the addition of a dependency to a completed future", () => {
```

Comment:
Note that this will be more strict in the future. Maybe add a todo?

Reply:
Can you elaborate how it will become more strict?

Reply:
You should only be able to add a dependency if it was successfully executed before the dependent started executing. That mimics how the execution without restarts works.